[1.18.x] Sound Providers: Electric Boogaloo (#438)
MinecraftForge/Documentation: docs/datagen/client/sounds.md

Sound Definition Generation
===========================

The `sounds.json` file can be generated for a mod by subclassing `SoundDefinitionsProvider` and implementing `#registerSounds`. After implementation, the provider must be [added][datagen] to the `DataGenerator`.

Adding a Sound
--------------

A sound definition can be generated by specifying the sound name and definition via `#add`. The sound name can either be provided from a [`SoundEvent`][soundevent], `ResourceLocation`, or string.

!!! warning
    The sound name supplied will always assume the namespace is the mod id supplied to the constructor of the provider. There is no validation performed on the namespace of the sound name!

### `SoundDefinition`

> The `SoundDefinition` can be supplied via `#definition`. The definition is responsible for containing the data to define a sound instance.

Reviewer: Instead of `supplied`, perhaps `created`? And we can remove the "is responsible for" phrase in the second sentence, for conciseness.
```suggestion
The `SoundDefinition` can be created using `#definition`. The definition contains the data to define a sound instance.
```

:---:      | :---
`with`     | Adds a sound(s) which may be played when the definition is selected.
`subtitle` | Sets the translation key of the definition.
`replace`  | When `true`, removes the sounds already defined by other `sounds.json` for this definition instead of appending to it.

### `SoundDefinition$Sound`

A sound supplied to the `SoundDefinition` can be specified using `SoundDefinitionsProvider#sound`. These methods take in the reference of the sound and a `SoundType` if specified.

The `SoundType` can be one of two values:

Sound Type | Definition
:---:      | :---
`SOUND`    | Specifies a reference to the sound located at `assets/<namespace>/sounds/<path>.ogg`.
`EVENT`    | Specifies a reference to the name of another sound defined by the `sounds.json`.

Each `Sound` created from `SoundDefinitionsProvider#sound` can specify additional configurations on how to load and play the sound provided:

Method | Description
:---: | :---
`volume` | Sets the volume scale of the sound, must be greater than `0`.
`pitch` | Sets the pitch scale of the sound, must be greater than `0`.
`weight` | Sets the likelihood of the sound getting played when the sound is selected.

> `stream` | When `true`, reads the sound from file instead of loading the sound into memory.

Reviewer: Perhaps this entry for `stream` should mention that this is recommended for long sounds, such as background music or music discs.

> Method | Description
> :---: | :---
> `volume` | Sets the volume scale of the sound, must be greater than `0`.
> `pitch` | Sets the pitch scale of the sound, must be greater than `0`.
> `weight` | Sets the likelihood of the sound getting played when the sound is selected.
> `stream` | When `true`, reads the sound from file instead of loading the sound into memory.
> `attenuationDistance` | Sets the number of blocks the sound can be heard from.
> `preload` | When `true`, immediately loads the sound into memory as soon as the resource pack is loaded.

Reviewer: Table column separators should be aligned.
```suggestion
Method                | Description
:---:                 | :---
`volume`              | Sets the volume scale of the sound, must be greater than `0`.
`pitch`               | Sets the pitch scale of the sound, must be greater than `0`.
`weight`              | Sets the likelihood of the sound getting played when the sound is selected.
`stream`              | When `true`, reads the sound from file instead of loading the sound into memory.
`attenuationDistance` | Sets the number of blocks the sound can be heard from.
`preload`             | When `true`, immediately loads the sound into memory as soon as the resource pack is loaded.
```
